stats/src/main/java/com/facebook/airlift/stats/QuantileDigest.java

```diff
@@ -1315,4 +1337,172 @@ public String toString()
 
         double middle(long lowerBound, long upperBound);
     }
+
+    public static class QuantileDigestIterator
+    {
+        private final int[] lefts; // left children
+        private final int[] rights; // right children
+        private final int[] stack; // array stack
+        private final double[] counts; // counts of nodes
+        private final byte[] levels; // levels in tree of nodes
+        private final long[] values; // values of nodes
+        private final double weightedCount; // total count of tree nodes
+        private final long min;
+        private final long max;
+        private final BitSet seen;
+
+        private int currentNode = -1; // current node in tree
+        private int stackPosition = -1; // current head position in stack array
+        private double cumulativeCount; // cumulative count of nodes in iteration
+        private double previousCumulativeCount; // previous value of cumulative count (0 if first element)
+        private boolean advanced; // iterator position state
+
+        public QuantileDigestIterator(int root, int[] lefts, int[] rights, double[] counts, byte[] levels, long[] values, double weightedCount, long min, long max)
+        {
+            this.weightedCount = weightedCount;
+            this.lefts = lefts;
+            this.rights = rights;
+            this.counts = counts;
+            this.levels = levels;
+            this.values = values;
+            this.min = min;
+            this.max = max;
+            seen = new BitSet(values.length);
+            stack = new int[(Integer.highestOneBit(values.length - 1) << 1) + 2]; // log2 ceiling
```

Comment:
nit: what's do you mean `ceiling(log2(values.length))` ?

Reply:
I don't understand the math 😂. What is the +2 for?

Reply:
The extra `+1` which makes this different from line 193:

`        int[] stack = new int[(Integer.highestOneBit(nodeCount - 1) << 1) + 1]; // log2 ceiling`

Is because in the iterator I include a `-1` sentinel.  I'll add a comment about it.

Reply:
Upon closer examination, this wasn't doing what anyone was expecting :)  The comment is just not accurate because it's more akin to snapping to the closest power of 2, which is way more space than needed.  At the risk of sounding obvious, one just needs a stack that is sized to the height of the tree.